version: '3.9'

services:
  api:
    container_name: api
    env_file: ${ENV_FILE}
    build:
      context: .
      dockerfile: Dockerfile
    ports: 
      - ${API_PORT}:${API_PORT}
    deploy:
      restart_policy:
        condition: on-failure
        delay: 10s
        max_attempts: 5
        window: 90s
    logging:
        driver: "json-file"
        options:
            max-file: 5
            max-size: 10m
    volumes: 
     - ./app:/app
     - ./nginx-unit.json:/var/lib/unit/conf.json # seems to be undocumented atm
     # - ./postgresql.key:/root/.postgresql/postgresql.key # total hack right here; ignore this in api/utils/db.py
    networks:
      - ergopad-data_net
      - ergopad-node_net
    # command: uvicorn main:app --workers 4 --host 0.0.0.0 --port ${API_PORT} --proxy-headers --use-colors --forwarded-allow-ips '52.72.64.235'
    # command: uvicorn main:app --reload --workers 2 --reload-dir /app --host 0.0.0.0 --port ${API_PORT} --proxy-headers --use-colors
    # command: gunicorn main:app --workers 4 --worker-class uvicorn.workers.UvicornWorker --bind 0.0.0.0:${API_PORT} --timeout 60
    # command: tail /dev/null -f
    # command: unitd-debug --no-daemon --control unix:/var/run/control.unit.sock
    command: unitd --no-daemon --control unix:/var/run/control.unit.sock

    # NOTE: 5/2022 Leif, using gunicorn seems to lock up on certain calls; could not find a configuration that worked; uvicorn does not have same issue
    # uvicorn.workers.UvicornH11Worker # use for pure python implementation
    # command: python -m gunicorn --bind :8000 --workers 4 main:app --worker-class uvicorn.workers.UvicornWorker --preload --timeout 0 --graceful-timeout 0 --worker-tmp-dir /dev/shm --keep-alive 2 

networks:
  ergopad-data_net:
    external: true
  ergopad-node_net:
    external: true
  net:
    driver: bridge
